### PR TITLE
Fixes #13135, ensure completude of the reactor summary but priviledge failures to be last to stay human efficient

### LINKED ISSUE #13135: Success module build are no more in summary when a module fails
### Affected version

4.0.0-rc-6

### Bug description

Follow up of https://github.com/apache/maven/pull/11977

Looked like a good idea but the report looks like only the failed modules were built whereas dozens more can have passed, idea is to sort the modules so failures are at the end so it stays readable but we have a complete summary

### PR DESCRIPTION
**Pull request by rmannibucau**
Goal is to ensure the reactor summary stays complete so we see some modules were built and not just the one(s) failling.

Following this checklist to help us incorporate your
contribution quickly and easily:

- [X] Your pull request should address just one issue, without pulling in other changes.
- [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [X] Each commit in the pull request should have a meaningful subject line and body.
  Note that commits might be squashed by a maintainer on merge.
- [X] Write unit tests that match behavioral changes, where the tests fail if the changes to the runtime are not applied.
  This may not always be possible but is a best-practice.
- [X] Run `mvn verify` to make sure basic checks pass.
  A more thorough check will be performed on your pull request automatically.
- [-] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

- [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
- [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by slawekjaranowski** [comment]
Skipping modules for which the build passed was intentional, according to issue
- https://github.com/apache/maven/issues/8027

Now, when we have a project with hundreds of modules, we will have a full list and need to scroll a long way up to see the root error.

Maybe this behavior should be parameterized?

**Comment by slawekjaranowski** [comment]
Another idea: show a summary with only the number of successful and skipped modules, and print the details only for the ones that failed.

For pass build we print all details as is now.

**Comment by rmannibucau** [comment]
@slawekjaranowski 

> Now, when we have a project with hundreds of modules, we will have a full list and need to scroll a long way up to see the root error.

not at all, this PR solves the original issue by sorting by status, failures last.

note that having a single module in the report for hundreds of module built is at least as much bothering than scrolling (looks like a bug in the execution and not only the report)

guess the proposal solves both cases and don't need a parameter since you don't need to scroll anymore (and colors help) and you have the failures immediately at the end (where you terminal is if it autoscroll like most of them)

on code feedback: will handle the hasException and buffer points

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java [hunk 1]
```diff
@@ -261,6 +282,12 @@ private void logReactorSummary(MavenSession session) {
         }
     }
 
+    private record ReactorSummaryRequest(
+            List<ReactorSummaryEntry> entries, StringBuilder buffer, boolean isSingleVersion, boolean hasExceptions) {}
```
[the remark this answers is not in the document]
Comment (rmannibucau) [reply compat/maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java:286]: this is a workaround for the aggressive checkstyle, will keep it until we relax checkstyle

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java [hunk 2]
```diff
@@ -199,31 +200,51 @@ private void logReactorSummary(MavenSession session) {
...
             } else if (buildSummary instanceof BuildFailure) {
                 statusMessage = failureMessage;
-                shouldSkip = false;
+                group = 2;
+            } else if (buildSummary == null) {
+                statusMessage = skippedMessage;
+                group = 0;
             } else {
                 statusMessage = unknownMessage;
+                group = 0;
```
[the remark this answers is not in the document]
Comment (rmannibucau) [reply compat/maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java:225]: acceptable I think for now

**Comment by slawekjaranowski** [comment]
> @slawekjaranowski
> 
> > Now, when we have a project with hundreds of modules, we will have a full list and need to scroll a long way up to see the root error.
> 
> not at all, this PR solves the original issue by sorting by status, failures last.
> 
> note that having a single module in the report for hundreds of module built is at least as much bothering than scrolling (looks like a bug in the execution and not only the report)
> 
> guess the proposal solves both cases and don't need a parameter since you don't need to scroll anymore (and colors help) and you have the failures immediately at the end (where you terminal is if it autoscroll like most of them)
> 
> on code feedback: will handle the hasException and buffer points

Not exactly, look at my screen shots

Maven 4.0.0-rc-6:

image

I see everything on one screen

Your proposition:

image

I need scroll terminal up to see description of errors

**Review by slawekjaranowski — CHANGES_REQUESTED** [review]
In original issue we have:

> Can we have a way to skip reactor logging, i.e. the reactor build order and/or reactor summary.
>
> I have hundreds of modules in my reactor build and I always have to scroll back pages to see the error message.

- https://github.com/apache/maven/issues/8027

here we remove requirements of skipping ... so issue should be reopened or closed as not planed if we merge this one.

**Comment by slawekjaranowski** [comment]
I am convinced that skipping correctly built modules when an error occurs is a good solution.
I would like to hear the opinions of others, including the author of the original issue - @delanym

for me -1 sorting/grouping introduce the problem again

**Comment by rmannibucau** [comment]
@slawekjaranowski ok I get your point but this is only relevant to a very particular case: monothreaded build, for all other cases it is counter productive. Also your two screenshots highlights the issue: "why did it build a single module? it is obviously why the test failed", so this is more misleading in most cases IMHO. I'd also like to emphasis we already had a solution to skip the summary: `-Dmaven.logger.log.org.apache.maven.cling.event.ExecutionEventLogger=warn` so adjusting the log level to the status of the build would make sense. can it be a compromise which will make everything happy?

side note: sorting solves the issue "where is the failed module" when you have hundreds of module (camel, hop, tomee, nifi etc), doesnt solve "present me all errors" but we'll never solve this one by design until it is a trivial case = single module with a single simple error just cause the errors can take more than a screen and you can get a chain of errors (think JAXRS for ex), so don't think we should fight the issue you mention but only the summary one.

**Comment by gnodet** [comment]
Thanks for the fix. Worth noting that #12697 (and its dependency chain starting at #12695) is introducing new event loggers (`PlainExecutionEventLogger`, `RichBuildEventListener`, etc.) that replace `ExecutionEventLogger`. The reactor summary behaviour you're fixing here would need to be ported to those new loggers as well — or alternatively, this fix could be implemented there directly if the new chain lands first. Just flagging so the work isn't duplicated or lost.

**Comment by rmannibucau** [comment]
@gnodet good call, guess this will need some agreement on the target now before we get the new ones :(. Right now I needed to revert several builds to rc5 due to that so hope we dont take too much time :crossed_fingers: .

**Comment by delanym** [comment]
@rmannibucau for a long running build `-Dmaven.logger.log.org.apache.maven.cling.event.ExecutionEventLogger=warn ` or `-Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.event.ExecutionEventLogger=warn` leaves you staring at a blank screen wondering whether its progressing or frozen. We do do that, and then get echo-plugin to print `[INFO] Building a:b:c`. Bit of a hack.

**Comment by rmannibucau** [comment]
@delanym the `org.apache.maven.cli.event.ExecutionEventLogger.detail` logger hack sounds like solving that and global build status stays on `org.apache.maven.cli.event.ExecutionEventLogger` logger. But looks better than current state where it looks like the build just didn't work (not failed, didnt work)

**Comment by gnodet** [comment]
> @gnodet good call, guess this will need some agreement on the target now before we get the new ones :(. Right now I needed to revert several builds to rc5 due to that so hope we dont take too much time 🤞 .

If there's a regression, we can fix it in rc7.  The PR I pointed at are all for 4.1.0.

**Comment by gnodet** [comment]
> In original issue we have:
> 
> > Can we have a way to skip reactor logging, i.e. the reactor build order and/or reactor summary.
> > I have hundreds of modules in my reactor build and I always have to scroll back pages to see the error message.
> 
> * #8027
> 
> here we remove requirements of skipping ... so issue should be reopened or closed as not planed if we merge this one.

But this issue does not advertise fixing #8027.  It just links to it because it's related IIUC.

**Comment by rmannibucau** [comment]
@gnodet my understand of the issue was that there are too much summary lines so you dont see the failed one fast enough, this part is solved by this PR. Slawomir completed it was also a goal to see the test failure inline immediately so solve _space_ -> this one I think will never really be solved by design until you do yet another "logger" (listener) which does filter and there we'll not get an agreement of the default so my proposal is to stay conservative on the default, keep a consistent output (all or nothing, sorted if it helps, but never a filtered summary which looks just like maven was broken for the run) and accept to scroll for test cause you will scroll anyway since the suite is executed in general (once again some edge cases would benefit from the code which led to the regression but these are corner cases).

**Comment by rmannibucau** [comment]
are we good now like that and we revisit with the new listeners - one being summary-free maybe?

**Comment by slawekjaranowski** [comment]
Never mind of technical implementation, we should listen users what they need. 
Here @delanym  need do some a hack to achieve expected result. 
For this change we have opinion only of two maintainer, maybe we can try simple discussion on user ML to better know wat users needs.

**Comment by rmannibucau** [comment]
@slawekjaranowski a very gently reminder I'm an user too and this is a bug for the user I am ;) - no technical concern there, if you want to do it using asm i'm fine

**Comment by delanym** [comment]
@slawekjaranowski it sounds like bigger and better things are coming, so that's just fine. I'm playing the long game.

**Comment by rmannibucau** [comment]
FTR: don't get me wrong, I'm not against enabling that use case (this was the logger code in the PR I removed after Guillaume's feedback), but I'm against an impacting regression. 100% aligned what Guillaume prepared should cover everyone needs so stability short term, enhancement mid terms.

**Review by gnodet — APPROVED** [review]

**Review by gnodet — APPROVED** [review]
I think this is the correct way to go until we can do more in 4.1.0.
The problem here, not having the info, is worse than having too much info.

Before merging, the technical aspect needs to be finalized, and the other issue closed as not planned anymore.

**Comment by rmannibucau** [comment]
@gnodet can you highlight the tech aspect you have in mind (happy to discuss on slack if it helps), latest report of claude review were not relevant from my point of view (or intended to be more exact) and the skipped line issue was fixed IIRC. About the other issue, do you reference https://github.com/apache/maven/issues/8027 ? Think we can keep open and see if the new listeners can fix it, no?

**Comment by gnodet** [comment]
> @gnodet can you highlight the tech aspect you have in mind (happy to discuss on slack if it helps), latest report of claude review were not relevant from my point of view (or intended to be more exact) and the skipped line issue was fixed IIRC. About the other issue, do you reference https://github.com/apache/maven/issues/8027 ? Think we can keep open and see if the new listeners can fix it, no?

That one seemed relevant, you disagree with the analysis ?

https://github.com/apache/maven/pull/13136#issuecomment-5682599917

**Comment by rmannibucau** [comment]
@gnodet (on the phone) thought i fixed this one but yeah I agree

**Comment by gnodet** [comment]
> @gnodet (on the phone) thought i fixed this one but yeah I agree

I'll push a fix.

**Comment by gnodet** [comment]
> > @gnodet (on the phone) thought i fixed this one but yeah I agree
> 
> I'll push a fix.

You actually fixed it, you were right. 

LGTM

**Review by slawekjaranowski — APPROVED** [review]
ok, in favor of new console modes ... I will not block it.

please be aware of similar change in 4.0.x and 3.10.x branches

some nit to consider or confirm that is intended: 
- order: skipped, success, failure - is different that old 
- with error log level - we have inconsistent indentation in the messages

**Comment by rmannibucau** [comment]
@slawekjaranowski ordering is intended to help identify failed modules faster (not the error but which ones which is already a baby step forward), indentation is a side effect. I think the benefit > the lost this way but on my side no strong issue to revert the level if desired.
